### PR TITLE
setting for enums by name/ordinal

### LINKED ISSUE #1197: ArgumentFactory for enums by ordinal
**Scenario**
Currently, during SQL statements building, JDBI translates enums to String using `.name()` method as we can see [here](https://github.com/jdbi/jdbi/blob/v3.3.0/core/src/main/java/org/jdbi/v3/core/argument/BuiltInArgumentFactory.java#L160). That can be replaced by extending `ArgumentFactory` class with a personalized behavior.

**Feature proposal**
In this issue, I suggest this default behavior to be decoupled in such a manner that we will have an `EnumByOrdinalArgumentFactory` (to be written) and an `EnumByNameArgumentFactory` (to be copied), with the latter as the default `ArgumentFactory` for enums, for the matter of compatibility.

By doing so, the developer may easily choose which `ArgumentFactory` he/she will use for mapping a hypothetical enum to their database. JDBI also gains a more decoupled approach.

### LINKED ISSUE #1273: Let user configure if enums should be mapped by name or ordinal
From #1268 

This config should also apply to the ArgumentFactory.

### LINKED ISSUE #1354: @ByName and @ByOrdinal for enum values
To override the mapping config locally

### PR DESCRIPTION
**Pull request by TheRealMarnes**
fixes #1273
fixes #1197 as far as I'm concerned (I don't think the factories are needed as API)
fixes #1354

### CODE REVIEW

File: core/src/main/java/org/jdbi/v3/core/Handles.java [hunk 1]
```diff
@@ -50,8 +50,9 @@ public boolean isForceEndTransactions() {
      * @param forceEndTransactions whether to enforce transaction termination
      *                             discipline.
      */
-    public void setForceEndTransactions(boolean forceEndTransactions) {
+    public Handles setForceEndTransactions(boolean forceEndTransactions) {
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/Handles.java]: This is a backwards incompatible change and cannot be merged.

File: core/src/main/java/org/jdbi/v3/core/argument/Arguments.java [hunk 2]
```diff
@@ -114,11 +117,12 @@ private Arguments register(QualifiedArgumentFactory factory) {
      * we don't have a type for.
      * @param untypedNullArgument the argument to bind
      */
-    public void setUntypedNullArgument(Argument untypedNullArgument) {
+    public Arguments setUntypedNullArgument(Argument untypedNullArgument) {
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/argument/Arguments.java]: Another non-permissible change

File: core/src/main/java/org/jdbi/v3/core/argument/EnumArgumentFactory.java [hunk 3]
```diff
@@ -15,17 +15,21 @@
...
 class EnumArgumentFactory implements ArgumentFactory {
     @Override
     public Optional<Argument> build(Type expectedType, Object rawValue, ConfigRegistry config) {
-        // Enums must be bound as VARCHAR
-        // TODO use the same configuration as EnumMapperFactory for consistency
         if (rawValue instanceof Enum) {
             Enum<?> enumValue = (Enum<?>) rawValue;
-            return Optional.of(new LoggableToStringOrNPEArgument<>(enumValue, Enum::name));
+
+            StatementBinder<Enum<?>> binder = config.get(Arguments.class).getBindEnumsByName()
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/argument/EnumArgumentFactory.java]: 👍

File: core/src/main/java/org/jdbi/v3/core/Handles.java (hunk 1 above)
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/Handles.java]: https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.4.15

Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/Handles.java]: I had such a feeling...

**Comment by TheRealMarnes** [comment]
Changes removed

File: core/src/main/java/org/jdbi/v3/core/config/Configurable.java [hunk 4]
```diff
@@ -100,6 +100,30 @@ default This setSqlParser(SqlParser parser) {
...
 
+    /**
+     * convenience method to set the binding and mapping policy for enums to use ordinal values.
+     *
+     * @return this
+     */
+    @SuppressWarnings("unchecked")
+    default This handleEnumsByOrdinal() {
+        getConfig(Arguments.class).setBindEnumsByName(false);
+        getConfig(ColumnMappers.class).setMapEnumsByName(false);
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/config/Configurable.java]: Feels like we want an `Enums` config class with a single property for this.

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory.java [hunk 5]
```diff
@@ -13,21 +13,21 @@
...
 import java.lang.reflect.Type;
 import java.util.Optional;
-
 import org.jdbi.v3.core.config.ConfigRegistry;
 
+import static org.jdbi.v3.core.generic.GenericTypes.getErasedType;
+
 /**
- * Produces enum column mappers, which map enums from numeric columns according to ordinal value.
+ * Produces enum column mappers, which map enums from numeric columns according to {@link Enum#ordinal()}.
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory.java]: Would be nice to add a `{@link}` to the new config property

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumMapperFactory.java [hunk 6]
```diff
@@ -22,9 +22,11 @@
  */
 class EnumMapperFactory implements ColumnMapperFactory {
     private static final EnumByNameMapperFactory BY_NAME = new EnumByNameMapperFactory();
+    private static final EnumByOrdinalMapperFactory BY_ORDINAL = new EnumByOrdinalMapperFactory();
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumMapperFactory.java]: Not a blocker for this PR, but what do you think of having `@ByName` and `@ByOrdinal` qualifying annotations which can override the preferred strategy on a case-by-case basis?

File: core/src/main/java/org/jdbi/v3/core/config/Configurable.java (hunk 4 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/config/Configurable.java]: Well I thought people might want to set them up to behave differently by default in read and write, like if they mainly read from one set of tables with a convention for one way, and mostly write to other tables with the other convention. That and I think `Enums` is even more of a dead-end config class than Time...

I'm 50/50 for both, doesn't really matter to me. Shall I change it?

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumMapperFactory.java (hunk 6 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/mapper/EnumMapperFactory.java]: Yeah, that was brought up before, I'm in favor of it. #1354

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory.java (hunk 5 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory.java]: After we decide on the location of said new preference

File: core/src/main/java/org/jdbi/v3/core/config/Configurable.java (hunk 4 above)
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/config/Configurable.java]: I also think it should be a simple property.  I don't think having read and write behavior differ makes all that much sense.

**Comment by TheRealMarnes** [comment]
Remarks handled, looks good to go to me

**Comment by TheRealMarnes** [comment]
Added a callout to Enums in the devguide.

**Comment by TheRealMarnes** [comment]
Also fixes #1197 as far as I'm concerned

**Comment by TheRealMarnes** [comment]
Working on a potential related improvement for these but I need a little time to figure it out, will remove the label when done/discarded

**Comment by TheRealMarnes** [comment]
To help with reviewing this:
- both binding and mapping are handled by 3 factories: an unqualified master that checks the `Enums` setting and composes a qualified type that it passed back to the chain, to 2 specialized factories that pass the enum in a converted form (name/ordinal) to the chain again for the actual binding
- the argument part was never exposed but the mapping part was made "API", so I just left it all there to be removed in the future as it's redundant and not needed for the user
- we kinda need #1363 to make this feature nicer to use in the fluent api

The diff is kinda big but I think it's easy to breeze through content-wise. I can break it up into more atomic commits if you want.

File: core/src/main/java/org/jdbi/v3/core/argument/EnumByNameArgumentFactory.java [hunk 7]
```diff
@@ -14,20 +14,26 @@
...
+import java.sql.Types;
 import java.util.Optional;
-import org.jdbi.v3.core.argument.internal.strategies.LoggableToStringOrNPEArgument;
+import org.jdbi.v3.core.EnumByName;
 import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.generic.GenericTypes;
 
-class EnumArgumentFactory implements ArgumentFactory {
+@EnumByName
+class EnumByNameArgumentFactory implements ArgumentFactory {
```
[the remark this answers is not in the document]
Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/argument/EnumByNameArgumentFactory.java]: In fact you could probably do without EnumUnqualifiedArgumentFactory class acting as a switchboard as follows:

```java
// after the enum type check
Set<Annotation> enumByName = singleton(EnumQualifiers.enumByName());
if (enumByName.equals(qualifiedType.getQualifiers())
    || enumByName.equals(Qualifiers.getQualifiers(enumClass))
    || config.get(Enums.class).enumsHandledByName()) {
  // this enum is bound by name
}
```

[the remark this answers is not in the document]
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/argument/EnumByNameArgumentFactory.java]: So you something like this?
image

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumMapper.java [hunk 8]
```diff
@@ -42,50 +39,7 @@
      * @return an enum mapper that matches on {@link Enum#ordinal()}
      */
     public static <E extends Enum<E>> ColumnMapper<E> byOrdinal(Class<E> type) {
-        return new ByOrdinal<>(type);
+        return new EnumByOrdinalMapper<>(type);
     }
 
-    private static class ByName<E extends Enum<E>> extends EnumMapper<E> {
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumMapper.java]: This was all nicely localized in one easily read file.  Now it's in five :(

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory2.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.Optional;
+import org.jdbi.v3.core.EnumByOrdinal;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.generic.GenericTypes;
+
+/**
+ * Produces enum column mappers, which map enums from numeric columns according to their {@link Enum#ordinal()}.
+ */
+@EnumByOrdinal
+class EnumByOrdinalMapperFactory2 implements ColumnMapperFactory {
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory2.java]: How come we duplicate all these mapper factories and deprecate the old ones for removal?  Can we just deprecate them "for hiding" and continue to use retrofitted existing implementations?  This is a very high LOC add for a small feature.

File: core/src/main/java/org/jdbi/v3/core/EnumConfig.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+ */
+package org.jdbi.v3.core;
+
+import org.jdbi.v3.core.config.JdbiConfig;
+
+/**
+ * Configuration for behavior related to {@link Enum}s.
+ */
+public class EnumConfig implements JdbiConfig<EnumConfig> {
+    private boolean handleEnumsByName = true;
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/EnumConfig.java]: Would it be useful to have this be a `private Annotation defaultEnumQualifier`, a more general mechanism that doesn't explicitly encode a choice between the two built in strategies?

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory2.java (hunk 9 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory2.java]: I duplicated them to avoid tangling/spaghetti. I deprecated the old ones for hiding, and when the time comes you can just hit `Delete` on the old ones and all will be right. I like to sever connections between new and old and leave as little shared space as possible, or at least extract the common parts to an independent space where it can be used by old and new without them touching each other. This way you can just forget about deprecated things and delete them at a glance, they take up physical space but have no bearing on things.

I've never found LOC to be a very meaningful measurement compared to things like coupling, scope size/code piling, or excessive abstractions.

I think we should skip this pr until later if there's this much food for arguments in it.

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumMapper.java (hunk 8 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/mapper/EnumMapper.java]: I'm not a fan of lobbing things together for non-technical reasons? Single responsibility, minimalism, yadda yadda?

File: core/src/main/java/org/jdbi/v3/core/EnumConfig.java (hunk 10 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/EnumConfig.java]: Are you talking about the API or internal workings? We can change the boolean internals to something tertiary when the need arises. The binary nature of the API did bother me too though. I was thinking an enum would be a less surprising thing for users to have to use than `setEnumHandling(EnumByName.class)`

Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/EnumConfig.java]: Actually, I take that back. `setEnumHandling(EnumByName.class)` is probably pretty neat, less redundant stuff for users to have to link together since they'll just be using the annotations as points of contact for everything to do with the feature.

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory2.java (hunk 9 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory2.java]: Would extracting ByName and ByOrdinal to independent package-private classes so they don't need to be duplicated satisfy you more?

File: core/src/main/java/org/jdbi/v3/core/EnumConfig.java (hunk 10 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/EnumConfig.java]: I think I would like a marker interface on those annotations then though, so you can't pass just any `Annotation` to the methods. That'd be like going back to java 5.

Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/EnumConfig.java]: We can't make that method value-safe though, users would be able to pass in just any annotation class. I'd go with an enum that collects the different annotations as values, only to be used for safety on the setter.

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumMapper.java (hunk 8 above)
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/mapper/EnumMapper.java]: I don't see this approach as particularly minimal, unfortunately.  I'm taking a pass at reducing some of the copy+paste.

File: core/src/main/java/org/jdbi/v3/core/EnumConfig.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        handling = other.handling;
+    }
+
+    /**
+     * Applies to both binding and mapping.
+     *
+     * @return true if enums are handled by name, false if enums are handled by ordinal
+     */
+    @Deprecated
+    public boolean isEnumHandledByName() {
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/EnumConfig.java]: This is adding already deprecated code?

File: core/src/main/java/org/jdbi/v3/core/EnumConfig.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import org.jdbi.v3.core.config.JdbiConfig;
+
+/**
+ * Configuration for behavior related to {@link Enum}s.
+ */
+public class EnumConfig implements JdbiConfig<EnumConfig> {
+    private static final Set<Class<? extends Annotation>> ALLOWED_HANDLINGS = new HashSet<>(Arrays.asList(EnumByName.class, EnumByOrdinal.class));
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/EnumConfig.java]: This restricts the added qualifying annotations, which removes the ability to bind your own handling.  Is there a specific reason to forbid custom handling?  The purpose of allowing the user to set their own choice was to allow e.g. vendor extensions if they need to bind it as a special type of object, and this forbids that..

Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/EnumConfig.java]: Maybe instead we should just encourage them to register their own handler as a higher priority instead? 🤷‍♀️

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/AdvancableClock.java [hunk 13]
```diff
@@ -16,27 +16,32 @@
 import java.time.Clock;
 import java.time.Instant;
 import java.time.ZoneId;
+import java.time.ZonedDateTime;
 import java.time.temporal.TemporalUnit;
 
-class MockClock extends Clock {
-    private Instant now = Instant.now();
+public class AdvancableClock extends Clock {
```
Comment (stevenschlansker) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/AdvancableClock.java]: How come this class name changed?  This seems unrelated to the meat of this PR.  It seems unrelated to time handling?

Comment (TheRealMarnes) [reply sqlobject/src/test/java/org/jdbi/v3/sqlobject/AdvancableClock.java]: Just a tiny collateral improvement, came across it somehow and thought it needed a spit shine. Instead of throwing exceptions and assuming system timezone data, it should just take up the timezone as part of its function. Turns it into a functional component instead of just a hack to get around something. I can take the change out if it bothers you but it's so small it may as well just be merged along ¯\\\_(ツ)_\/¯

File: core/src/main/java/org/jdbi/v3/core/argument/EnumByNameArgumentFactory.java [hunk 14]
```diff
@@ -14,21 +14,25 @@
...
 import java.util.Optional;
-
-import org.jdbi.v3.core.argument.internal.strategies.LoggableBinderArgument;
+import org.jdbi.v3.core.EnumByName;
 import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.internal.Enums;
 
-class EnumArgumentFactory implements ArgumentFactory {
+@EnumByName
+class EnumByNameArgumentFactory implements ArgumentFactory {
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/argument/EnumByNameArgumentFactory.java]: I think it would be much simpler to change this class (which is package-private, so this would not break compatibility) to implement `QualifiedArgumentFactory` instead of `ArgumentFactory`.

The precedence rules could live in e.g. `EnumConfig.resolveEnumStrategy(QualifiedType)`, which would return the correct `EnumStrategy` enum value based on the precedence rules:

```java
public EnumStrategy resolveEnumStrategy(QualifiedType type) {
  EnumStrategy strategy = resolveEnumStrategy(type.getQualifiers());

  if (strategy == null) {
    Class<?> enumType = getErasedType(type.getType());
    strategy = resolveEnumStrategy(getQualifiers(enumType));
  }

  if (strategy == null) {
    strategy = defaultStrategy;
  }
  
  return strategy;
}

private EnumStrategy resolveEnumStrategy(Collection<Annotation> qualifiers) {
  if (qualifiers.contains(EnumByName.class)) {
    return EnumStrategy.BY_NAME;
  }

  if (qualifiers.contains(EnumByOrdinal.class)) {
    return EnumStrategy.BY_ORDINAL;
  }

  return null;
}
```

Then the argument factory and mapper factory can delegate to that config class to resolve the enum strategy, and just pick the existing argument / mapper implementation according to the resolved strategy.

File: core/src/main/java/org/jdbi/v3/core/argument/EnumByOrdinalArgumentFactory.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.lang.reflect.Type;
+import java.sql.Types;
+import java.util.Optional;
+import org.jdbi.v3.core.EnumByOrdinal;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.internal.Enums;
+
+@EnumByOrdinal
+class EnumByOrdinalArgumentFactory implements ArgumentFactory {
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/argument/EnumByOrdinalArgumentFactory.java]: Note that this factory becomes redundant if my suggestion to implement `QualifiedArgumentFactory` is used.

File: core/src/main/java/org/jdbi/v3/core/argument/EnumUnqualifiedArgumentFactory.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.lang.reflect.Type;
+import java.util.Optional;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.internal.Enums;
+
+/**
+ * catches any non-qualified enums, qualifies them, and passes them back to the chain
+ */
+class EnumUnqualifiedArgumentFactory implements ArgumentFactory {
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/argument/EnumUnqualifiedArgumentFactory.java]: And this class would also be redundant.

File: core/src/main/java/org/jdbi/v3/core/internal/Enums.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.Set;
+import org.jdbi.v3.core.EnumByName;
+import org.jdbi.v3.core.EnumByOrdinal;
+import org.jdbi.v3.core.EnumConfig;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.generic.GenericTypes;
+import org.jdbi.v3.core.qualifier.QualifiedType;
+import org.jdbi.v3.core.qualifier.Qualifiers;
+
+public class Enums {
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/internal/Enums.java]: @TheRealMarnes Just a note on design philosophy of config classes that I'm not sure we've communicated to you:

We've been naming most config classes with pluralized names so that inline usage with the `Configurable.configure()` method reads as "configure row mappers", "configure arguments", etc. So I think I'd like to rename `EnumConfig` to `Enums`. Thus:

```
handle.createUpdate(sql)
    .configure(Enums.class, config -> config.setDefaultEnumStrategy(EnumStrategy.BY_ORDINAL))
    .execute();
```

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByNameMapperFactory2.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.Optional;
+import org.jdbi.v3.core.EnumByName;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.generic.GenericTypes;
+
+/**
+ * Produces enum column mappers, which case-insensitively map enums from textual columns according to their {@link Enum#name()}.
+ */
+@EnumByName
+class EnumByNameMapperFactory2 implements ColumnMapperFactory {
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumByNameMapperFactory2.java]: I think we'd be better served by refactoring `EnumMapperFactory` (which is package private) to implement `QualifiedColumnMapperFactory`, and follow the same logic I suggested for `EnumArgumentFactory`. Then these new classes are redundant.

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory2.java (hunk 9 above)
Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapperFactory2.java]: This class likewise becomes redundant if `EnumMapperFactory` implements `QualifiedColumnMapperFactory`

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByNameMapperFactory.java [hunk 19]
```diff
@@ -22,7 +22,11 @@
 
 /**
  * Produces enum column mappers, which map enums from varchar columns using {@link Enum#valueOf(Class, String)}.
+ *
+ * @deprecated there is no reason for this to be API
  */
+@Deprecated
+// TODO jdbi4: delete
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumByNameMapperFactory.java]: The deprecation message is inaccurate. This and `EnumByOrdinalMapperFactory` were made public API to enable changing the default enum strategy via a registration:

Core:

```
jdbi.registerColumnMapper(EnumByOrdinalMapperFactory.class);
```

SQL Object:

```
@RegisterColumnMapperFactory(EnumByOrdinalMapperFactory.class)
```

With that said, we never added an equivalent for `@RegisterArgumentFactory`, so it wasn't a full solution.

Let's change the deprecation message to redirect users to the new API options that replace this class: the config class `default*` property, or the annotations.

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapper.java [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    EnumByOrdinalMapper(Class<E> type) {
+        this.constants = type.getEnumConstants();
+        this.type = type;
+    }
+
+    @Override
+    public E map(ResultSet rs, int columnNumber, StatementContext ctx) throws SQLException {
+        Integer ordinal = ctx.findColumnMapperFor(Integer.class)
+            .orElseThrow(() -> new UnableToProduceResultException("an Integer column mapper is required to map Enums from ordinals", ctx))
+            .map(rs, columnNumber, ctx);
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapper.java]: Feels like overkill looking up the mapper from the registry when we can just call `getInt()`. The mappers for core java types are very unlikely to be overridden.

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByNameMapper.java [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+    private final ConcurrentMap<String, E> lookups = new ConcurrentHashMap<>();
+
+    EnumByNameMapper(Class<E> type) {
+        this.type = type;
+    }
+
+    @Override
+    public E map(ResultSet rs, int columnNumber, StatementContext ctx) throws SQLException {
+        String name = ctx.findColumnMapperFor(String.class)
+            .orElseThrow(() -> new UnableToProduceResultException("a String column mapper is required to map Enums from names", ctx))
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumByNameMapper.java]: Ditto my comment in `EnumByOrdinalMapper.map()`

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapper.java (hunk 20 above)
Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/mapper/EnumByOrdinalMapper.java]: If you feel strongly about keeping this, I'd vote move this lookup to `specialize()` so it's not repeated on every row.

File: core/src/main/java/org/jdbi/v3/core/mapper/EnumMapper.java [hunk 22]
```diff
@@ -13,17 +13,14 @@
...
-
 /**
  * Column mapper for Java {@code enum} types.
  * @param <E> the enum type mapped
+ *
+ * @deprecated there is no reason for this to be API
  */
+@Deprecated
+// TODO jdbi4: delete
 public abstract class EnumMapper<E extends Enum<E>> implements ColumnMapper<E> {
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/EnumMapper.java]: This class is still the only access point for `EnumMapper.byName()` and `EnumMapper.byOrdinal()`, e.g. for use with `ResultBearing.map(ColumnMapper<T>): ResultIterable<T>`

File: core/src/test/java/org/jdbi/v3/core/mapper/TestEnums.java [hunk 23]
```diff
@@ -83,7 +84,7 @@ public void testMapInvalidEnumValue() {
 
         assertThatThrownBy(() -> h.createQuery("select * from something order by id")
             .mapToBean(SomethingElse.class)
-            .findFirst()).isInstanceOf(IllegalArgumentException.class);
+            .findFirst()).isInstanceOf(UnableToProduceResultException.class);
```
Comment (qualidafial) [inline core/src/test/java/org/jdbi/v3/core/mapper/TestEnums.java]: Why did this change?

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/EnumByXTest.java [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+
+            FooDao dao = h.attach(FooDao.class);
+
+            dao.insert(Foo.BAR);
+            assertThat(h.createQuery("select ordinal from enums").mapTo(Integer.class).findOnly()).isEqualTo(0);
+
+            Foo value = dao.select();
+            assertThat(value).isEqualTo(Foo.BAR);
+        });
+    }
```
Comment (qualidafial) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/EnumByXTest.java]: Probably also want to check `@EnumByName` when the default is set to by-ordinal.

**Comment by qualidafial** [comment]
Those are all my comments for now.

**Comment by TheRealMarnes** [comment]
closing to merge locally and polish up for a fresh pr with a cleaner slate

File: core/src/test/java/org/jdbi/v3/core/mapper/TestEnums.java (hunk 23 above)
Comment (TheRealMarnes) [reply core/src/test/java/org/jdbi/v3/core/mapper/TestEnums.java]: Because the new implementations throw this exception instead of a general IAE
